src/TerraformInitCommand.groovy

```diff
@@ -44,8 +55,12 @@ class TerraformInitCommand {
         if (!input) {
             pieces << "-input=false"
         }
-        backendConfigs.each { config ->
-            pieces << "-backend-config=${config}"
+        if(doBackend) {
+            backendConfigs.each { config ->
+                pieces << "-backend-config=${config}"
+            }
+        } else {
+            pieces << "-backend=false"
```

Comment:
See comment

Reply:
Given a provider configuration of `terraform { backend { s3 { } }`, a plain `terraform init -input=false` will fail claiming that you have not provided a region, bucket, or key.  Since almost all the backend configs are environment dependent, and the validate stage is effectively without an environment, `-backend=false` is required to _skip_ the config and avoid these errors.  backendConfigs has to be skipped because your backend plugins are going still write those arguments to that array.

We aren't undoing a previously initialized backend, we are skipping the backend.

> To skip backend configuration, use `-backend=false`. Note that some other init steps require an initialized backend, so it is recommended to use this flag only when the working directory was already previously initialized for a particular backend.

I'll work on reworking it to `withoutBackend()` pattern.

Reply:
I read your answer, but was previously still not fully understanding.  I've since installed 0.12.8, and played around with it, and better understand your change now.  The piece that I was missing was:

1) Prior to 0.12.*, you could run `terraform validate` without first running `terraform init`
2) As of 0.12.+, you are forced to run `terraform init` prior to being able to run `terraform validate`

I'm mulling over the change that you're making - adding `terraform init` to the TerraformValidateStage, then setting `-backend=false`.   In an ideal world, I would push the addition of `terraform init` into TerraformPlugin as well - that is, prior to 0.12, `terraform init` would not be called as part of TerraformValidateStage, but 0.12+ would have the command added.  

TerraformValidateStage is not currently that flexible.

I'm not ignoring this PR, just mulling over the consequences.  I'd be willing to accept this PR into the v5.3-dev branch, but I think I'd want to continue building on your PR before publishing an official release.

Reply:
I get it.  In the long term, you want the plugin to be the source of truth for how things need to be adjusted between versions.  Currently the only real way without a deep rework of Stages and their interface the best we can do is keep adding switches and have the plugin flip those switches.  Which while better than the original implementation, still spreads out the logic of 0.11 vs 0.12 vs future versions across a lot of different classes.

I've got a fork I'm running off of in the meantime, so take your time.  It's a needly problem. especially with the way the terraform devs seem willing to completely rework the workflow for various ... goals so to speak, as 0.12 showed.

Outside of adding `runBefore()` and `runAfter()` stages to every single command, what happens in the future when 1.0 decides validate is just "plan" with "-check-variables=false -refresh-state=false"?!?  It seems like we are approaching a world where the plugin would need to pick and register a Stage Implementation and at that point are we over engineering a solution to a problem that hasn't yet occured?

Having typed all this out, I think pre and post hooks on the Stages might just be the best answer until a terraform version throws us a new wrench that doesn't solve.

Reply:
^---  Yeah, pre/post hooks were exactly what I had in mind.  A pre/post hook around JUST `terraform validate`, would give `TerraformPlugin` the ability to conditionally inject a `terraform init` based on the version being used.

I'll take one last look at the changes, merge this change in, note this discussion on Issue #109 as remaining work before closing it out, then I'll attack it with a PR when  I get the chance.

Thank you again!